### PR TITLE
[LoongArch] Support amcas[_db].{b/h/w/d} instructions.

### PR DESCRIPTION
**Pull request by tangaac**
Two options for clang: -mlamcas & -mno-lamcas.
Enable or disable amcas[_db].{b/h} instructions.
The default is -mno-lamcas.
Only works on LoongArch64.

### CODE REVIEW

File: llvm/test/CodeGen/LoongArch/ir-instruction/atomicrmw-lamcas.ll [hunk 1]
```diff
@@ -0,0 +1,5025 @@
...
+; LA64-LAMCAS-NEXT:    ext.w.b $a4, $a0
+; LA64-LAMCAS-NEXT:    amcas.b $a0, $a3, $a2
+; LA64-LAMCAS-NEXT:    bne $a0, $a4, .LBB128_1
+; LA64-LAMCAS-NEXT:  # %bb.2: # %atomicrmw.end
+; LA64-LAMCAS-NEXT:    ret
+  %1 = atomicrmw xor ptr %a, i8 %b monotonic
+  ret i8 %1
+}
+
+define i16 @atomicrmw_xor_i16_monotonic(ptr %a, i16 %b) nounwind {
```
Comment (tangaac) [inline llvm/test/CodeGen/LoongArch/ir-instruction/atomicrmw-lamcas.ll:4998]: When -mlamcas is enabled, it shouldn't use amcas for atomicrmw_(and\or\xor)_(i8\i16).
﻿
It could amand.w, amor.w amxor.w to expand.
https://github.com/llvm/llvm-project/blob/5df84a75351d0e9c3e20d50ac1047c937e3b8e88/llvm/lib/CodeGen/AtomicExpandPass.cpp#L670C1-L684C4

How to fix it?

File: clang/test/Driver/loongarch-mlamcas.c [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+// CC1-LAMCAS: "-target-feature" "+lamcas"
+// CC1-NO-LAMCAS: "-target-feature" "-lamcas"
+
+// IR-LAMCAS: attributes #[[#]] ={{.*}}"target-features"="{{(.*,)?}}+lamcas{{(,.*)?}}"
+// IR-NO-LAMCAS: attributes #[[#]] ={{.*}}"target-features"="{{(.*,)?}}-lamcas{{(,.*)?}}"
+
+int foo(void) {
+  return 42;
+}
```
Comment (xen0n) [inline clang/test/Driver/loongarch-mlamcas.c:30]: missing trailing newline

File: llvm/lib/Target/LoongArch/AsmParser/LoongArchAsmParser.cpp [hunk 3]
```diff
@@ -1562,7 +1562,9 @@ unsigned LoongArchAsmParser::checkTargetMatchPredicate(MCInst &Inst) {
   unsigned Opc = Inst.getOpcode();
   switch (Opc) {
   default:
-    if (Opc >= LoongArch::AMADD_D && Opc <= LoongArch::AMXOR_W) {
+    // amcas[_db].{b/h/w/d} didn't need this judgement
+    if ((Opc >= LoongArch::AMADD_B && Opc <= LoongArch::AMAND__DB_W) ||
+        (Opc >= LoongArch::AMMAX_D && Opc <= LoongArch::AMXOR__DB_W)) {
```
Comment (xen0n) [inline llvm/lib/Target/LoongArch/AsmParser/LoongArchAsmParser.cpp:1567]: Ah. This is very prone to breakage as correct operation relies on the order of insn definitions in the TableGen files.

For now it may be okay to merge as-is, but let me try refactoring this piece of logic to hopefully make it elegant and future-proof.

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp [hunk 4]
```diff
@@ -363,6 +363,10 @@ LoongArchTargetLowering::LoongArchTargetLowering(const TargetMachine &TM,
   setPrefFunctionAlignment(Subtarget.getPrefFunctionAlignment());
   setPrefLoopAlignment(Subtarget.getPrefLoopAlignment());
   setMaxBytesForAlignment(Subtarget.getMaxBytesForAlignment());
+
+  // Set MinCmpXchg size for 'LAMCAS' feature.
```
Comment (xen0n) [inline llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:367]: This is stating the obvious (the reader would know this from the code alone), let's give more background:

```suggestion
  // cmpxchg sizes down to 8 bits become legal if LAMCAS is available.
```

(Or we can actually delete the comment altogether, as this is really obvious even without the explanation. Anyway either is better than re-stating the *what* instead of *why*.)

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp [hunk 5]
```diff
@@ -5743,6 +5747,14 @@ LoongArchTargetLowering::shouldExpandAtomicRMWInIR(AtomicRMWInst *AI) const {
   }
 
   unsigned Size = AI->getType()->getPrimitiveSizeInBits();
+  if (Subtarget.hasLAMCAS()) {
+    if (AI->getOperation() == AtomicRMWInst::Nand ||
+        (Size < 32 /*&& (AI->getOperation() != AtomicRMWInst::Or &&
```
Comment (xen0n) [inline llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:5752]: How is this related to code size?

**Review by xen0n — COMMENTED** [review]
I think this is getting more convoluted than necessary? For example, I didn't see a `setOperationAction(ISD::ATOMIC_CMP_SWAP, Legal)` call gated on presence of the `lamcas` feature, which I think would render a lot (if not most) of the changes here made to the various `shouldXxx` predicates unnecessary.

**Comment by xen0n** [comment]
Also please split the LLVM and Clang changes.

File: llvm/lib/Target/LoongArch/AsmParser/LoongArchAsmParser.cpp (hunk 3 above)
Comment (tangaac) [reply llvm/lib/Target/LoongArch/AsmParser/LoongArchAsmParser.cpp:1567]: > Ah. This is very prone to breakage as correct operation relies on the order of insn definitions in the TableGen files.

I think so.

After v1.1 instructions added, AMADD_B has the smallest opc because of dictionary order.
While AMXOR__DB_W already has the largest opc, not AMXOR_W. The previous code should be a bug.

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp (hunk 5 above)
Comment (tangaac) [reply llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:5752]: atomicrmw nand & i8 / i16 atomicrmw add / sub / [u]min / [u]max should be converted to amcas.

i8 / i16 atomicrmw and  / or / xor shouldn't be converted to amcas because they can use am(and / or / xor).w to expand. 

But after feature LAMCAS enabled, all i8,i16 atomicrmw operations will be converted to amcas, 
how to fix it?

File: llvm/test/CodeGen/LoongArch/ir-instruction/atomicrmw-lamcas.ll [hunk 6]
```diff
@@ -0,0 +1,5025 @@
...
+; LA64-NEXT:    srl.w $a0, $a4, $a2
+; LA64-NEXT:    ret
+;
+; LA64-LAMCAS-LABEL: atomicrmw_xchg_i8_acquire:
+; LA64-LAMCAS:       # %bb.0:
+; LA64-LAMCAS-NEXT:    ld.bu $a2, $a0, 0
+; LA64-LAMCAS-NEXT:    .p2align 4, , 16
+; LA64-LAMCAS-NEXT:  .LBB0_1: # %atomicrmw.start
+; LA64-LAMCAS-NEXT:    # =>This Inner Loop Header: Depth=1
+; LA64-LAMCAS-NEXT:    ext.w.b $a3, $a2
```
Comment (heiher) [inline llvm/test/CodeGen/LoongArch/ir-instruction/atomicrmw-lamcas.ll:32]: Can we eliminate this sign extension?

Comment (tangaac) [reply llvm/test/CodeGen/LoongArch/ir-instruction/atomicrmw-lamcas.ll:32]: No, the reason for this sign extension is because ld.bu.
All i8, i16 use ld.bu / ld.hu to load the value, and use  ext.w.b / ext.w.h  later
i32, i64 use ld.w / ld.d to load the value, and use move later
So the number of insts is the same.

```llvm
; LA64-LAMCAS-LABEL: atomicrmw_nand_i8_acquire:
; LA64-LAMCAS:       # %bb.0:
; LA64-LAMCAS-NEXT:    move $a2, $a0
; LA64-LAMCAS-NEXT:    ld.bu $a0, $a0, 0
; LA64-LAMCAS-NEXT:    .p2align 4, , 16
; LA64-LAMCAS-NEXT:  .LBB18_1: # %atomicrmw.start
; LA64-LAMCAS-NEXT:    # =>This Inner Loop Header: Depth=1
; LA64-LAMCAS-NEXT:    and $a3, $a0, $a1
; LA64-LAMCAS-NEXT:    nor $a3, $a3, $zero
; LA64-LAMCAS-NEXT:    ext.w.b $a4, $a0
; LA64-LAMCAS-NEXT:    amcas_db.b $a0, $a3, $a2
; LA64-LAMCAS-NEXT:    bne $a0, $a4, .LBB18_1
; LA64-LAMCAS-NEXT:  # %bb.2: # %atomicrmw.end
; LA64-LAMCAS-NEXT:    ret

; LA64-LAMCAS-LABEL: atomicrmw_nand_i32_acquire:
; LA64-LAMCAS:       # %bb.0:
; LA64-LAMCAS-NEXT:    move $a2, $a0
; LA64-LAMCAS-NEXT:    ld.w $a0, $a0, 0
; LA64-LAMCAS-NEXT:    .p2align 4, , 16
; LA64-LAMCAS-NEXT:  .LBB20_1: # %atomicrmw.start
; LA64-LAMCAS-NEXT:    # =>This Inner Loop Header: Depth=1
; LA64-LAMCAS-NEXT:    move $a3, $a0
; LA64-LAMCAS-NEXT:    and $a4, $a0, $a1
; LA64-LAMCAS-NEXT:    nor $a4, $a4, $zero
; LA64-LAMCAS-NEXT:    amcas_db.w $a0, $a4, $a2
; LA64-LAMCAS-NEXT:    bne $a0, $a3, .LBB20_1
; LA64-LAMCAS-NEXT:  # %bb.2: # %atomicrmw.end
; LA64-LAMCAS-NEXT:    ret
```

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp [hunk 7]
```diff
@@ -5722,6 +5726,58 @@ bool LoongArchTargetLowering::getTgtMemIntrinsic(IntrinsicInfo &Info,
   }
 }
 
+void LoongArchTargetLowering::emitExpandAtomicRMW(AtomicRMWInst *AI) const {
```
Comment (SixWeining) [inline llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:5729]: Please add some comments for this func and add a test (ir -> ir).

File: llvm/test/Transforms/AtomicExpand/LoongArch/atomicrmw-expand.ll
Comment (SixWeining) [inline llvm/test/Transforms/AtomicExpand/LoongArch/atomicrmw-expand.ll:1]: Just pick one memory ordering is enough as this test is not sensitive to memory ordering.

**Review by SixWeining — APPROVED** [review]

**Review by heiher — APPROVED** [review]

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp [hunk 8]
```diff
@@ -5768,6 +5828,15 @@ LoongArchTargetLowering::shouldExpandAtomicRMWInIR(AtomicRMWInst *AI) const {
   }
 
   unsigned Size = AI->getType()->getPrimitiveSizeInBits();
+  if (Subtarget.hasLAMCAS()) {
+    if (Size < 32 && (AI->getOperation() == AtomicRMWInst::And ||
+                      AI->getOperation() == AtomicRMWInst::Or ||
+                      AI->getOperation() == AtomicRMWInst::Xor))
+      return AtomicExpansionKind::Expand;
+    if ((AI->getOperation() == AtomicRMWInst::Nand || Size < 32))
```
Comment (SixWeining) [inline llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:5836]:
```suggestion
    if (AI->getOperation() == AtomicRMWInst::Nand || Size < 32)
```

File: llvm/lib/Target/LoongArch/LoongArchInstrInfo.td [hunk 9]
```diff
@@ -2119,6 +2119,26 @@ def : Pat<(atomic_load_sub_i16 GPR:$rj, GPR:$rk),
           (AMADD__DB_H (SUB_W R0, GPR:$rk), GPR:$rj)>;
 } // Predicates = [ IsLA64, HasLAM_BH ]
 
+let Predicates = [ HasLAMCAS, IsLA64 ] in {
+
```
Comment (SixWeining) [inline llvm/lib/Target/LoongArch/LoongArchInstrInfo.td:2123]: useless blank line

**Review by SixWeining — APPROVED** [review]
